src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -625,7 +625,24 @@ public static unsafe bool TryCompleteConnect(SafeSocketHandle socket, int socket
             Interop.Error err;
             try
             {
-                err = Interop.Sys.GetSocketErrorOption(socket, &socketError);
+                // Due to fd recyling, TryCompleteConnect may be called when there was a write event
+                // for the previous socket that used the fd.
+                // The SocketErrorOption in that case is the same as for a succesful connect.
+                // To filter out these false events, we check whether the socket is writable, before
+                // reading the socket option.
+                Interop.PollEvents outEvents;
```

Comment:
Any chance we can define tests for FD recycling?

Reply:
FD recyling is common on Unix platforms. FDs don't increment (as I think is the case on Windows). syscalls that return fds will usually return the lowest free fd.